### PR TITLE
fix(qwik): process loops before show; correct state merging

### PR DESCRIPTION
**Pull request by mhevery**
## Description

- `For` loops need to be processed before `Show`/`Hide` as `Show`/`Hide` can refer to `$index$` from the `For` loop.
- The state merging of component `state`, `props` and `For`-loop `locals` have be improved to be consistent with React implementation.

### CODE REVIEW

File: packages/core/src/helpers/nodes/for.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ export const getForArguments = (
 ): string[] => {
   return [
     node.scope.forName || 'item',
-    node.scope.indexName,
+    node.scope.indexName || '$index',
```
Comment (samijaber) [inline packages/core/src/helpers/nodes/for.ts:10]: Why is this change needed for the fixes? Is `$index` a special var name that needs to stay as-is?

Comment (mhevery) [reply packages/core/src/helpers/nodes/for.ts:10]: Yes, that is my understanding. 

Ohh, thinking more; maybe that is for BuilderApp only?

**Review by samijaber — APPROVED** [review]
